File: 2019/16/16.py
signal = []
BASE = [0, 1, 0, -1]
BASE_CACHE = []

with open('input.txt', 'rt') as textfile:
    for line in textfile:
        line_i = int(line)
        while line_i is not 0:
            signal.insert(0, line_i % 10)
            line_i //= 10

def p_print(l):
    output = ""
    for element in l:
        output += str(element)
    return int(output)

def rel_base(n):
    if n < len(BASE_CACHE):
        return BASE_CACHE[n]
    output = []
    for item in BASE:
        for i in range(n+1):
            output.append(item)
    BASE_CACHE.append(output)
    return output

def rep_base(n, l):
    base = rel_base(n)
    output = []
    while len(output) < (l + 1):
        output += base
    output.pop(0)
    return output

def fft(start, phases):
    old = start.copy()
    for i in range(phases):
        new = []
        for j in range(len(old)):
            base = rep_base(j, len(old))
            new.append(abs(sum([old[i] * base[i] for i in range(len(old))])) % 10)
        old = new.copy()
    return new

def ffft(start, phases):
    start.reverse()
    old = start.copy()
    for i in range(phases):
        new = []
        total = 0
        for j in range(len(old)):
            total += old[j]
            new.append(total % 10)
        old = new.copy()
    new.reverse()
    return new

print(p_print(fft(signal, 100)[0:8]))
length = len(signal)
offset = p_print(signal[0:7])
signal *= 10000
signal = signal[offset - (offset % length):]
offset -= offset - (offset % length)
print(p_print(ffft(signal, 100)[offset:offset + 8]))

File: 2019/16/16.jl
input = readlines("input.txt")[1]
array = [parse(Int64, char) for char in input]
offset = sum([array[k]*10^(7-k) for k=1:7])

const BASE = [0, 1, 0, -1]

function pattern(pos, el)
	if pos <= el % 4pos < 2pos
		return 1
	elseif 3pos <= el % 4pos
		return -1
	else
		return 0
	end
end

function fft(array)
	new = zeros(Int64, length(array))
	Threads.@threads for i in 1:length(array)
		new[i] = abs(sum([pattern(i, j) * array[j] for j in 1:length(array)])) % 10
	end
	return new
end

function ffft(array)
	new = zeros(Int64, length(array))
	total = 0
	for i in length(array):-1:1
		total += array[i]
		new[i] = total % 10
	end
	return new
end

function p1(array, iters)
	for _ in 1:iters
		array .= fft(array)
	end
	return array
end

function p2(array, iters)
	for _ in 1:iters
		array .= ffft(array)
	end
	return array
end

println(p1(copy(array), 100)[1:8])
println(p2(repeat(array, 10000)[offset+1:end], 100)[1:8])
